File: backend/src/infrastructure/services/certificate.service.ts
import PDFDocument from 'pdfkit';
import QRCode from 'qrcode';
import fs from 'fs';
import path from 'path';
import { env } from '../../config/env';
import { v4 as uuidv4 } from 'uuid';

interface CertificateData {
  certificateNo: string;
  applicationNo: string;
  serviceName: string;
  citizenName: string;
  issuedDate: Date;
  verificationUrl: string;
}

export class CertificateService {
  async generateCertificate(data: CertificateData): Promise<{ filePath: string; qrCode: string }> {
    const qrCodeData = await QRCode.toDataURL(data.verificationUrl);
    const fileName = `certificate-${data.certificateNo}.pdf`;
    const filePath = path.join(env.UPLOAD_DIR, fileName);

    return new Promise((resolve, reject) => {
      const doc = new PDFDocument({ size: 'A4', margin: 50 });
      const stream = fs.createWriteStream(filePath);

      doc.pipe(stream);

      doc.fontSize(24).fillColor('#1e40af').text('Government of India', { align: 'center' });
      doc.moveDown(0.5);
      doc.fontSize(18).text('Official Certificate', { align: 'center' });
      doc.moveDown(2);

      doc.fontSize(12).fillColor('#000');
      doc.text(`Certificate No: ${data.certificateNo}`);
      doc.text(`Application No: ${data.applicationNo}`);
      doc.text(`Service: ${data.serviceName}`);
      doc.text(`Issued To: ${data.citizenName}`);
      doc.text(`Date of Issue: ${data.issuedDate.toLocaleDateString('en-IN')}`);
      doc.moveDown(2);

      doc.text('This is to certify that the above-mentioned application has been duly processed and approved by the competent authority.', {
        align: 'justify',
      });
      doc.moveDown(2);

      const qrBuffer = Buffer.from(qrCodeData.split(',')[1], 'base64');
      doc.image(qrBuffer, doc.page.width - 150, doc.y, { width: 100 });
      doc.fontSize(8).text('Scan to verify', doc.page.width - 150, doc.y + 105, { width: 100, align: 'center' });

      doc.end();

      stream.on('finish', () => resolve({ filePath, qrCode: qrCodeData }));
      stream.on('error', reject);
    });
  }

  generateCertificateNo(): string {
    const year = new Date().getFullYear();
    const random = uuidv4().split('-')[0].toUpperCase();
    return `CERT-${year}-${random}`;
  }
}

export const certificateService = new CertificateService();

File: backend-csharp/Services/CertificateService.cs
using QRCoder;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;

namespace GovernmentCitizenServices.Api.Services
{
    public class CertificateService
    {
        private readonly IConfiguration _configuration;

        public CertificateService(IConfiguration configuration)
        {
            _configuration = configuration;
            QuestPDF.Settings.License = LicenseType.Community;
        }

        public string GenerateCertificateNo()
        {
            var year = DateTime.UtcNow.Year;
            var random = Guid.NewGuid().ToString().Substring(0, 8).ToUpper();
            return $"CERT-{year}-{random}";
        }

        public (string filePath, string qrCodeBase64) GenerateCertificate(
            string certificateNo,
            string applicationNo,
            string serviceName,
            string citizenName,
            DateTime issuedDate,
            string verificationUrl)
        {
            var uploadDir = _configuration["UploadSettings:UploadDirectory"] ?? "./uploads";
            if (!Directory.Exists(uploadDir))
            {
                Directory.CreateDirectory(uploadDir);
            }

            var fileName = $"certificate-{certificateNo}.pdf";
            var filePath = Path.Combine(uploadDir, fileName);

            // Generate QR Code Base64
            using var qrGenerator = new QRCodeGenerator();
            using var qrData = qrGenerator.CreateQrCode(verificationUrl, QRCodeGenerator.ECCLevel.Q);
            using var qrCode = new PngByteQRCode(qrData);
            var qrBytes = qrCode.GetGraphic(20);
            var qrCodeBase64 = $"data:image/png;base64,{Convert.ToBase64String(qrBytes)}";

            // Generate PDF Document using QuestPDF
            Document.Create(container =>
            {
                container.Page(page =>
                {
                    page.Size(PageSizes.A4);
                    page.Margin(50);
                    page.PageColor(Colors.White);
                    page.DefaultTextStyle(x => x.FontSize(12).FontFamily("Arial"));

                    page.Header().Column(col =>
                    {
                        col.Item().AlignCenter().Text("Government of India").FontSize(24).Bold().FontColor(Colors.Blue.Medium);
                        col.Item().AlignCenter().Text("Official Certificate").FontSize(18).SemiBold().FontColor(Colors.Grey.Darken2);
                    });

                    page.Content().PaddingVertical(20).Column(col =>
                    {
                        col.Item().Text($"Certificate No: {certificateNo}").Bold();
                        col.Item().Text($"Application No: {applicationNo}");
                        col.Item().Text($"Service Name: {serviceName}");
                        col.Item().Text($"Issued To: {citizenName}");
                        col.Item().Text($"Date of Issue: {issuedDate:dd/MM/yyyy}");

                        col.Item().PaddingTop(20).Text(
                            "This is to certify that the above-mentioned application has been duly processed and approved by the competent authority."
                        ).Justify();

                        col.Item().PaddingTop(30).AlignRight().Width(120).Column(qrCol =>
                        {
                            qrCol.Item().Image(qrBytes);
                            qrCol.Item().AlignCenter().Text("Scan to verify").FontSize(8);
                        });
                    });

                    page.Footer().AlignCenter().Text("Official Digital Document — Verified Online").FontSize(9).FontColor(Colors.Grey.Medium);
                });
            }).GeneratePdf(filePath);

            return (filePath, qrCodeBase64);
        }
    }
}
